File: packages/schemas/mcp-tool-catalog.ts
import { z } from "zod"

export const mcpDiscoveryRunStatuses = ["succeeded", "failed"] as const
export const mcpDiscoveredToolAvailabilities = ["present", "removed"] as const
export const mcpDiscoveredToolDriftStatuses = [
  "new",
  "unchanged",
  "changed",
  "removed",
] as const

const jsonObjectSchema = z.record(z.string(), z.unknown())

export const mcpDiscoveryRunResponseSchema = z.object({
  id: z.string(),
  tenantId: z.string(),
  projectId: z.string(),
  connectionId: z.string(),
  status: z.enum(mcpDiscoveryRunStatuses),
  toolCount: z.number(),
  catalogFingerprint: z.string().nullable(),
  errorCode: z.string().nullable(),
  startedAt: z.string(),
  completedAt: z.string(),
  createdBy: z.string(),
})

export const mcpDiscoveredToolResponseSchema = z.object({
  id: z.string(),
  tenantId: z.string(),
  projectId: z.string(),
  connectionId: z.string(),
  name: z.string(),
  title: z.string().nullable(),
  description: z.string(),
  inputSchema: jsonObjectSchema,
  annotations: jsonObjectSchema,
  fingerprint: z.string(),
  enabled: z.boolean(),
  availability: z.enum(mcpDiscoveredToolAvailabilities),
  driftStatus: z.enum(mcpDiscoveredToolDriftStatuses),
  firstSeenAt: z.string(),
  lastSeenAt: z.string(),
  removedAt: z.string().nullable(),
  discoveryRunId: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
})

export const mcpToolCatalogResponseSchema = z.object({
  connectionId: z.string(),
  tools: z.array(mcpDiscoveredToolResponseSchema),
  latestRun: mcpDiscoveryRunResponseSchema.nullable(),
  lastSuccessfulRun: mcpDiscoveryRunResponseSchema.nullable(),
})

export const setMCPToolEnabledSchema = z.object({
  enabled: z.boolean(),
})

export type SetMCPToolEnabledSchemaProps = z.infer<
  typeof setMCPToolEnabledSchema
>

File: apps/api/internal/domain/entities/mcp_tool_catalog.go
package entities

import (
	"encoding/json"
	"time"
)

// MCPDiscoveryRunStatus describes the outcome of one explicit tools/list refresh.
type MCPDiscoveryRunStatus string

const (
	MCPDiscoverySucceeded MCPDiscoveryRunStatus = "succeeded"
	MCPDiscoveryFailed    MCPDiscoveryRunStatus = "failed"
)

// MCPDiscoveredToolAvailability indicates whether the provider returned a tool
// in the latest successful catalog.
type MCPDiscoveredToolAvailability string

const (
	MCPToolPresent MCPDiscoveredToolAvailability = "present"
	MCPToolRemoved MCPDiscoveredToolAvailability = "removed"
)

// MCPDiscoveredToolDriftStatus records how the tool changed in the latest
// successful refresh. It is informational and never rewrites authored bindings.
type MCPDiscoveredToolDriftStatus string

const (
	MCPToolNew          MCPDiscoveredToolDriftStatus = "new"
	MCPToolUnchanged    MCPDiscoveredToolDriftStatus = "unchanged"
	MCPToolChanged      MCPDiscoveredToolDriftStatus = "changed"
	MCPToolDriftRemoved MCPDiscoveredToolDriftStatus = "removed"
)

type MCPDiscoveryRun struct {
	ID                 string
	TenantID           string
	ProjectID          string
	ConnectionID       string
	Status             MCPDiscoveryRunStatus
	ToolCount          int
	CatalogFingerprint *string
	ErrorCode          *string
	StartedAt          time.Time
	CompletedAt        time.Time
	CreatedBy          string
}

// MCPDiscoveredTool is sanitized provider metadata. It deliberately contains
// no endpoint, headers, credentials, or executable information.
type MCPDiscoveredTool struct {
	ID             string
	TenantID       string
	ProjectID      string
	ConnectionID   string
	Name           string
	Title          *string
	Description    string
	InputSchema    json.RawMessage
	Annotations    json.RawMessage
	Fingerprint    string
	Enabled        bool
	Availability   MCPDiscoveredToolAvailability
	DriftStatus    MCPDiscoveredToolDriftStatus
	FirstSeenAt    time.Time
	LastSeenAt     time.Time
	RemovedAt      *time.Time
	DiscoveryRunID *string
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

type MCPToolCatalog struct {
	ConnectionID      string
	Tools             []MCPDiscoveredTool
	LatestRun         *MCPDiscoveryRun
	LastSuccessfulRun *MCPDiscoveryRun
}
